repo2docker/buildpacks/base.py

```diff
@@ -481,13 +482,38 @@ def render(self):
             labels=self.get_labels(),
             build_script_directives=build_script_directives,
             assemble_script_directives=assemble_script_directives,
-            build_script_files=self.get_build_script_files(),
+            build_script_files={
+                self.generate_build_context_filename(k)[0]: v
+                for k, v in self.get_build_script_files().items()
+            },
             base_packages=sorted(self.get_base_packages()),
             post_build_scripts=self.get_post_build_scripts(),
             start_script=self.get_start_script(),
             appendix=self.appendix,
         )
 
+    @staticmethod
+    def generate_build_context_filename(src_path):
+        """
+        Generate a filename for a file injected into Docker build context.
+
+        In case the src_path is relative, it's assumed it's relative to directory of
+        this __file__. Returns the resulting filename and an absolute path to the source
+        file on host.
+        """
+        if not os.path.isabs(src_path):
+            src_parts = src_path.split('/')
+            src_path = os.path.join(os.path.dirname(__file__), *src_parts)
+        BLOCKSIZE = 65536
+        hash_md5 = hashlib.md5()
+        hash_md5.update(src_path.encode())
+        with open(src_path, 'rb') as fh:
+            buf = fh.read(BLOCKSIZE)
+            while len(buf) > 0:
+                hash_md5.update(buf)
+                buf = fh.read(BLOCKSIZE)
+        return hash_md5.hexdigest(), src_path
```

Comment:
We should probably include at least some part of the filename in the resulting filepath we return, to make debugging easier. How about something like: `assemble_files/<escaped-file-path-truncated>-<6-chars-of-hash>` or something like that?

See https://github.com/jupyterhub/binderhub/blob/6b2908d7aaf4a7ec62beed0019de54db06494214/binderhub/builder.py#L127 where we do something like that, keeping in mind the 255 char linux filepath limit.

Reply:
Fixed in 5447508